File: lib/wType/boolean.cpp
#include "boolean.h"

W::Boolean::Boolean():
    Object(),
    data(false)
{

}

W::Boolean::Boolean(bool value):
    Object(),
    data(value)
{

}

W::Boolean::Boolean(const W::Boolean& original):
    Object(),
    data(original.data)
{

}

W::Boolean::~Boolean()
{

}

W::Boolean& W::Boolean::operator=(const W::Boolean& original)
{
    data = original.data;
    return *this;
}

W::Boolean& W::Boolean::operator=(bool original)
{
    data = original;
    return *this;
}

W::Object::StdStr W::Boolean::toString() const
{
    StdStr str;
    if (data)
    {
        str = "true";
    }
    else
    {
        str = "false";
    }
    return str;
}

W::Object::objectType W::Boolean::getType() const
{
    return Boolean::type;
}

W::Object::size_type W::Boolean::size() const
{
    return 1;
}

W::Object* W::Boolean::copy() const
{
    return new Boolean(*this);
}

bool W::Boolean::operator>(const W::Boolean& other) const
{
    return data > other.data;
}

bool W::Boolean::operator<(const W::Boolean& other) const
{
    return data < other.data;
}

bool W::Boolean::operator==(const W::Boolean& other) const
{
    return data == other.data;
}

bool W::Boolean::operator!=(const W::Boolean& other) const
{
    return data != other.data;
}

bool W::Boolean::operator<=(const W::Boolean& other) const
{
    return data <= other.data;
}

bool W::Boolean::operator>=(const W::Boolean& other) const
{
    return data >= other.data;
}

void W::Boolean::serialize(W::ByteArray& out) const
{
    uint8_t val;
    if (data)
    {
        val = 253;
    }
    else
    {
        val = 0;
    }
    
    out.push(val);
}

void W::Boolean::deserialize(W::ByteArray& in)
{
    uint8_t val = in.pop();
    
    if (val == 253)
    {
        data = true;
    }
    else
    {
        data = false;
    }
}

W::Boolean::operator bool() const
{
    return data;
}

File: lorgar/lib/wType/boolean.js
"use strict";
(function boolean_js() {
    var moduleName = "lib/wType/boolean";
    
    var defineArray = [];
    defineArray.push("lib/wType/object");
    
    define(moduleName, defineArray, function boolean_module() {
        var Object = require("lib/wType/object");
        
        var Boolean = Object.inherit({
            "className": "Boolean",
            "constructor": function(bool) {
                Object.fn.constructor.call(this);
                
                this._data = bool === true;
            },
            "<": function(other) {
                if (!(other instanceof Boolean)) {
                    throw new Error("Can compare Boolean only with Boolean");
                }
                return this._data < other._data;
            },
            ">": function(other) {
                if (!(other instanceof Boolean)) {
                    throw new Error("Can compare Boolean only with Boolean");
                }
                return this._data > other._data;
            },
            "==": function(other) {
                if (!(other instanceof Boolean)) {
                    throw new Error("Can compare Boolean only with Boolean");
                }
                return this._data === other._data;
            },
            "clone": function() {
                return new Boolean(this._data);
            },
            "deserialize": function(ba) {
                var int = ba.pop();
                
                if (int === 253) {
                    this._data = true;
                } else {
                    this._data = false;
                }
            },
            "serialize": function(ba) {
                if (this._data) {
                    ba.push(253);
                } else {
                    ba.push(0);
                }
            },
            "toString": function() {
                return this._data.toString();
            },
            "valueOf": function() {
                return this._data;
            }
        });
        
        return Boolean;
    });
})();